### PR TITLE
Copy & paste activity

### PR DESCRIPTION
**Pull request by pmattmann**
Copy and Paste Activities

### CODE REVIEW

**Comment by BacLuc** [comment]
Es scheint als wäre der clipboard zugriff auf firefox bisschen kompliziert:
```
TypeError: 'clipboard-read' (value of 'name' member of PermissionDescriptor) is not a valid value for enumeration PermissionName.
    mounted CopyActivityInfoDialog.vue:49
    VueJS 19
    click CampProgram.vue:81
    VueJS 4
    onClick LockButton.vue:60
    VueJS 33
```
Das passiert nur schon, wenn man den edit mode im picasso aktiviert.
Wenn man dann das Kopier symbol klickt, dann gibt es kein Feedback.

Beim erstellen einer neuen Aktivität wird die console dann auch rot.

Sobald man in der aktivitäts ansicht das three dot menu aufmacht, wird die konsole auch rot.

**Comment by BacLuc** [comment]
In Safari hat das navigator.permissions objekt gar keine properties (oder ist wahrscheinlich null).
Es gibt einen type error

File: api/src/Entity/Activity.php [hunk 1]
```diff
@@ -120,6 +120,13 @@ class Activity extends BaseEntity implements BelongsToCampInterface {
     #[ORM\JoinColumn(nullable: false)]
     public ?Category $category = null;
 
+    /**
+     * Copy Contents from this Source-Activity.
+     */
+    #[ApiProperty(example: '/activities/1a2b3c4d')]
+    #[Groups(['create'])]
```
Comment (BacLuc) [inline api/src/Entity/Activity.php:124]: Ich würde testen, dass das transient field nicht wieder ausgegeben wird nach dem create.
Sonst #[ApiProperty(readable: false)]

File: api/src/State/ActivityCreateProcessor.php [hunk 2]
```diff
@@ -26,16 +26,21 @@ public function __construct(
...
-
         if (!isset($data->category?->rootContentNode)) {
             throw new \UnexpectedValueException('Property rootContentNode of provided category is null. Object of type '.ColumnLayout::class.' expected.');
         }
-
         if (!is_a($data->category->rootContentNode, ColumnLayout::class)) {
             throw new \UnexpectedValueException('Property rootContentNode of provided category is of wrong type. Object of type '.ColumnLayout::class.' expected.');
         }
 
+        $data->camp = $data->category->camp;
```
Comment (BacLuc) [inline api/src/State/ActivityCreateProcessor.php:36]: hier sagst du, dass der null check von vorher unnötig war?

File: api/tests/Api/Activities/CreateActivityTest.php [hunk 3]
```diff
@@ -477,6 +477,71 @@ public function testCreateActivityValidatesMissingScheduleEntries() {
...
+            ['json' => $this->getExampleWritePayload(
+                [
+                    'copyActivitySource' => $this->getIriFor('activity1'),
+                    'category' => $this->getIriFor('category1'),
+                ],
+                []
+            )]
+        );
+
+        // No Access on activity1 -> BadRequest
```
Comment (BacLuc) [inline api/tests/Api/Activities/CreateActivityTest.php]: der response code sagt was anderes ;p

File: api/tests/Api/Activities/CreateActivityTest.php [hunk 4]
```diff
@@ -477,6 +477,71 @@ public function testCreateActivityValidatesMissingScheduleEntries() {
...
+                            'start' => '2023-03-25T15:00:00+00:00',
+                            'end' => '2023-03-25T16:00:00+00:00',
+                        ],
+                    ],
+                ],
+                []
+            )]
+        );
+
+        // No Access on activity1 -> BadRequest
```
Comment (BacLuc) [inline api/tests/Api/Activities/CreateActivityTest.php]: der response code sagt was anderes ;p

File: api/src/Entity/Activity.php [hunk 5]
```diff
@@ -120,6 +120,13 @@ class Activity extends BaseEntity implements BelongsToCampInterface {
     #[ORM\JoinColumn(nullable: false)]
     public ?Category $category = null;
 
+    /**
+     * Copy Contents from this Source-Activity.
```
Comment (BacLuc) [inline api/src/Entity/Activity.php:121]: Does the source start with a capital letter

File: api/tests/Api/Activities/CreateActivityTest.php [hunk 6]
```diff
@@ -509,6 +574,7 @@ public function getExampleWritePayload($attributes = [], $except = []) {
             Activity::class,
             Post::class,
             array_merge([
+                'copyActivitySource' => null,
```
Comment (BacLuc) [inline api/tests/Api/Activities/CreateActivityTest.php:577]: would it also work without this additional property?

File: api/tests/State/ActivityCreateProcessorTest.php [hunk 7]
```diff
@@ -41,6 +41,7 @@ protected function setUp(): void {
         $categoryRoot->contentType = $contentType;
         $this->activity->category->setRootContentNode($categoryRoot);
 
+        // EntityManager
```
Comment (BacLuc) [inline api/tests/State/ActivityCreateProcessorTest.php:44]: Looks like a repository to me

File: frontend/src/components/activity/CopyActivityInfoDialog.vue [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    :cancel-label="$tc('global.button.close')"
+  >
+    <template #activator="scope">
+      <slot name="activator" v-bind="scope" />
+    </template>
+
+    <p>
+      {{ $tc('components.activity.copyActivityInfoDialog.description') }}
+    </p>
+    <p v-if="clipboardReadState == 'prompt'">
```
Comment (BacLuc) [inline frontend/src/components/activity/CopyActivityInfoDialog.vue]: ===

File: frontend/src/components/activity/CopyActivityInfoDialog.vue [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+      <center>
+        <v-btn color="success" @click="requestClipboardAccess">
+          {{ $tc('components.activity.copyActivityInfoDialog.allow') }}
+        </v-btn>
+      </center>
+    </p>
+    <p v-if="clipboardReadState == 'granted'">
```
Comment (BacLuc) [inline frontend/src/components/activity/CopyActivityInfoDialog.vue]: ===

File: frontend/src/components/activity/CopyActivityInfoDialog.vue [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+      {{ $tc('components.activity.copyActivityInfoDialog.granted') }}
+    </p>
+    <p v-if="clipboardReadState == 'denied'">
```
Comment (BacLuc) [inline frontend/src/components/activity/CopyActivityInfoDialog.vue]: ===

File: frontend/src/components/activity/CopyActivityInfoDialog.vue [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+export default {
+  name: 'CopyActivityInfoDialog',
+  components: {
+    DialogForm,
+  },
+  extends: DialogBase,
+  data() {
+    return {
+      clipboardReadState: 'unknown',
```
Comment (BacLuc) [inline frontend/src/components/activity/CopyActivityInfoDialog.vue:44]: null is not good enough... or undefined?

File: frontend/src/components/activity/CopyActivityInfoDialog.vue [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    }
+  },
+  async mounted() {
+    // read current permission
+    let res = await navigator.permissions.query({ name: 'clipboard-read' })
```
Comment (BacLuc) [inline frontend/src/components/activity/CopyActivityInfoDialog.vue]: const

File: frontend/src/components/activity/CopyActivityInfoDialog.vue [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+    this.clipboardReadState = res.state
+  },
+  methods: {
+    cancel() {
+      this.close()
+    },
+    async requestClipboardAccess() {
+      // if permission is not yet requested, request it
+      if (this.clipboardReadState == 'prompt') {
```
Comment (BacLuc) [inline frontend/src/components/activity/CopyActivityInfoDialog.vue]: ===

File: frontend/src/components/activity/CopyActivityInfoDialog.vue [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+        try {
+          await navigator.clipboard.readText()
+        } catch {
+          console.log('clipboard read is denied')
+        }
+
+        let res = await navigator.permissions.query({ name: 'clipboard-read' })
```
Comment (BacLuc) [inline frontend/src/components/activity/CopyActivityInfoDialog.vue]: const

File: frontend/src/components/activity/ScheduleEntry.vue [hunk 15]
```diff
@@ -348,6 +370,23 @@ export default {
         this.editActivityTitle = true
       }
     },
+    async copyUrlToClipboard() {
+      let res = await navigator.permissions.query({ name: 'clipboard-read' })
```
Comment (BacLuc) [inline frontend/src/components/activity/ScheduleEntry.vue]: const

Comment (BacLuc) [inline frontend/src/components/activity/ScheduleEntry.vue]: const

File: frontend/src/components/activity/ScheduleEntry.vue [hunk 16]
```diff
@@ -348,6 +370,23 @@ export default {
...
+      if (res.state == 'prompt') {
+        this.$refs.copyInfoDialog.open()
+      }
+
+      let scheduleEntry = scheduleEntryRoute(this.scheduleEntry())
+      let url = window.location.origin + router.resolve(scheduleEntry).href
```
Comment (BacLuc) [inline frontend/src/components/activity/ScheduleEntry.vue]: const

File: frontend/src/components/program/DialogActivityCreate.vue [hunk 17]
```diff
@@ -41,14 +97,41 @@ export default {
...
+        return this.entityData.copyActivitySource != null
+      },
+      set(val) {
+        if (val) {
+          this.entityData.copyActivitySource = this.copyActivitySource._meta.self
+          this.entityData.title = this.copyActivitySource.title
+          if (
+            this.period().camp()._meta.self == this.copyActivitySource.camp()._meta.self
+          ) {
+            console.log(this.copyActivitySource.category()._meta.self)
```
Comment (BacLuc) [inline frontend/src/components/program/DialogActivityCreate.vue]: pls delete the console.log

File: frontend/src/components/program/DialogActivityCreate.vue [hunk 18]
```diff
@@ -61,15 +144,51 @@ export default {
...
+        this.$set(this, 'clipboardPermission', p.state)
+        this.$set(this, 'copyActivitySource', null)
+
+        if (p.state == 'granted') {
+          this.getCopyActivitySource().then((activity) => {
+            this.$set(this, 'copyActivitySource', activity)
+          })
+        }
+      })
+      console.log('refreshCopyActivitySource')
```
Comment (BacLuc) [inline frontend/src/components/program/DialogActivityCreate.vue]: pls remove the console log

File: frontend/src/components/program/DialogActivityCreate.vue [hunk 19]
```diff
@@ -41,14 +97,41 @@ export default {
   },
   data() {
     return {
+      clipboardPermission: 'unknown',
```
Comment (BacLuc) [inline frontend/src/components/program/DialogActivityCreate.vue:134]: null or undefined is not good enough?

File: frontend/src/components/program/DialogActivityCreate.vue [hunk 20]
```diff
@@ -61,15 +144,51 @@ export default {
               deleted: false,
             },
           ],
+          copyActivitySource: null,
         })
       } else {
+        this.canReadClipboard = 'unknown'
```
Comment (BacLuc) [inline frontend/src/components/program/DialogActivityCreate.vue:208]: null or undefined?

File: api/src/State/ActivityCreateProcessor.php (hunk 2 above)
Comment (pmattmann) [reply api/src/State/ActivityCreateProcessor.php:36]: Nein, vorher war die Abfrage vor dem `$data->category->rootContentNode`

File: api/tests/Api/Activities/CreateActivityTest.php (hunk 3 above)
Comment (pmattmann) [reply api/tests/Api/Activities/CreateActivityTest.php]: Ne, der ResponseCode ist `400 Bad Request`

File: api/tests/Api/Activities/CreateActivityTest.php (hunk 6 above)
Comment (pmattmann) [reply api/tests/Api/Activities/CreateActivityTest.php:577]: nop; `getExampleReadPayload()` does set it with a defaultvalue

File: api/tests/State/ActivityCreateProcessorTest.php (hunk 7 above)
Comment (pmattmann) [reply api/tests/State/ActivityCreateProcessorTest.php:44]: it does prepare the EntityManager

File: frontend/src/components/activity/CopyActivityInfoDialog.vue (hunk 11 above)
Comment (pmattmann) [reply frontend/src/components/activity/CopyActivityInfoDialog.vue:44]: if displayed somewhere (debug); 'unknown' more understandable

File: frontend/src/components/activity/CopyActivityInfoDialog.vue (hunk 12 above)
Comment (pmattmann) [reply frontend/src/components/activity/CopyActivityInfoDialog.vue]: thx

File: api/tests/Api/Activities/CreateActivityTest.php [hunk 21]
```diff
@@ -477,6 +477,71 @@ public function testCreateActivityValidatesMissingScheduleEntries() {
...
+                [
+                    'copyActivitySource' => $this->getIriFor('activity1'),
+                    'category' => $this->getIriFor('category1'),
+                ],
+                []
+            )]
+        );
+
+        // No Access on activity1 -> BadRequest
+        $this->assertResponseStatusCodeSame(201);
```
Comment (BacLuc) [inline api/tests/Api/Activities/CreateActivityTest.php:518]: der sagt aber 201

**Comment by DeNic0la** [comment]
Hey, I played around a bit and noticed some things.

Copying a block has a delay or does sometimes not work.

https://github.com/ecamp/ecamp3/assets/57986114/0f89b2d9-eb05-49fb-820c-a26227f8387b

The URL was copied to my clipboard (I pasted it into a new tab and it worked) but was not recognized by ecamp at first.

---

When deleting a block while in the copy process, there is no error message. (or it didn't load on my Browser)

https://github.com/ecamp/ecamp3/assets/57986114/5496972b-27eb-4879-87a7-1e8b538e76c6

Doing this is basically asking for trouble but the same thing happens when the internet connection is lost:
image

File: frontend/src/components/activity/ScheduleEntry.vue [hunk 22]
```diff
@@ -348,6 +370,27 @@ export default {
...
+        if (res.state == 'prompt') {
+          this.$refs.copyInfoDialog.open()
+        }
+      } catch {
+        console.warn('clipboard permission not requestable')
+      }
+
+      const scheduleEntry = scheduleEntryRoute(this.scheduleEntry())
+      const url = window.location.origin + router.resolve(scheduleEntry).href
+      navigator.clipboard.writeText(url)
```
Comment (DeNic0la) [inline frontend/src/components/activity/ScheduleEntry.vue]: This ensures the toast is only shown after successfully copying the text.
```suggestion
      await navigator.clipboard.writeText(url)
```
I think this might fix the first case I pointed out.

File: frontend/src/components/activity/ScheduleEntry.vue [hunk 23]
```diff
@@ -283,6 +298,13 @@ export default {
     scheduleEntries() {
       return this.activity.scheduleEntries().items
     },
+    activityName() {
+      return (
```
Comment (BacLuc) [inline frontend/src/components/activity/ScheduleEntry.vue:333]: maybe we already have this logic somewhere.
can be refactored in another PR.

File: frontend/src/components/program/DialogActivityCreate.vue [hunk 24]
```diff
@@ -16,21 +16,107 @@
...
+            <e-checkbox
+              v-model="copyContent"
+              :label="$tc('components.program.dialogActivityCreate.copyActivityContent')"
+            />
+          </v-list-item-action>
+        </v-list-item>
+      </div>
+    </div>
+    <dialog-activity-form :activity="entityData" :period="period">
+      <template v-if="clipboardUnsccessable" #textFieldTitleAppend>
```
Comment (BacLuc) [inline frontend/src/components/program/DialogActivityCreate.vue]: maybe clipboardAccessDenied?

File: frontend/src/components/program/DialogActivityCreate.vue [hunk 25]
```diff
@@ -61,15 +195,94 @@ export default {
...
+        }
+      )
+    },
+    async getCopyActivitySource(url) {
+      if (url?.startsWith(window.location.origin)) {
+        url = url.substring(window.location.origin.length)
+        let match = router.matcher.match(url)
+
+        if (match.name == 'activity') {
+          var scheduleEntry = await this.api
```
Comment (BacLuc) [inline frontend/src/components/program/DialogActivityCreate.vue]:
```suggestion
          const scheduleEntry = await this.api
```

File: frontend/src/components/program/DialogActivityCreate.vue [hunk 26]
```diff
@@ -61,15 +195,94 @@ export default {
...
+        () => {
+          this.clipboardPermission = 'unaccessable'
+          console.warn('clipboard permission not requestable')
+        }
+      )
+    },
+    async getCopyActivitySource(url) {
+      if (url?.startsWith(window.location.origin)) {
+        url = url.substring(window.location.origin.length)
+        let match = router.matcher.match(url)
```
Comment (BacLuc) [inline frontend/src/components/program/DialogActivityCreate.vue]:
```suggestion
        const match = router.matcher.match(url)
```

File: frontend/src/components/program/picasso/PicassoEntry.vue [hunk 27]
```diff
@@ -252,6 +269,26 @@ export default {
         this.scrollHeight = this.$el.scrollHeight
       })
     },
+    async copyUrlToClipboard() {
+      try {
+        let res = await navigator.permissions.query({ name: 'clipboard-read' })
```
Comment (BacLuc) [inline frontend/src/components/program/picasso/PicassoEntry.vue]:
```suggestion
        const res = await navigator.permissions.query({ name: 'clipboard-read' })
```

File: frontend/src/components/program/picasso/PicassoEntry.vue [hunk 28]
```diff
@@ -252,6 +269,26 @@ export default {
...
+        if (res.state == 'prompt') {
+          this.$refs.copyInfoDialog.open()
+        }
+      } catch {
+        console.warn('clipboard permission not requestable')
+      }
+
+      let url = window.location.origin + router.resolve(this.scheduleEntryRoute).href
```
Comment (BacLuc) [inline frontend/src/components/program/picasso/PicassoEntry.vue]:
```suggestion
      const url = window.location.origin + router.resolve(this.scheduleEntryRoute).href
```

File: api/src/Entity/Activity.php (hunk 1 above)
Comment (pmattmann) [reply api/src/Entity/Activity.php:124]: field is not returned

File: frontend/src/components/program/DialogActivityCreate.vue [hunk 29]
```diff
@@ -16,21 +16,107 @@
...
+          </v-list-item-action>
+        </v-list-item>
+      </div>
+    </div>
+    <dialog-activity-form :activity="entityData" :period="period">
+      <template v-if="clipboardAccessDenied" #textFieldTitleAppend>
+        <PopoverPrompt
+          v-model="copyActivitySourceUrlShowPopover"
+          icon="mdi-content-paste"
+          title="Kopierte Aktivität einfügen"
```
Comment (BacLuc) [inline frontend/src/components/program/DialogActivityCreate.vue]: This should be translated

File: frontend/src/components/program/DialogActivityCreate.vue [hunk 30]
```diff
@@ -16,21 +16,107 @@
...
+              v-on="scope.on"
+            >
+              <v-progress-circular v-if="copyActivitySourceUrlLoading" indeterminate />
+              <v-icon v-else>mdi-content-paste</v-icon>
+            </v-btn>
+          </template>
+          {{ $tc('components.program.dialogActivityCreate.copySourceInfo') }}
+          <e-text-field
+            v-model="copyActivitySourceUrl"
+            label="Aktivität kopieren"
```
Comment (BacLuc) [inline frontend/src/components/program/DialogActivityCreate.vue]: This should be translated

File: frontend/src/components/activity/ScheduleEntry.vue (hunk 23 above)
Comment (manuelmeister) [reply frontend/src/components/activity/ScheduleEntry.vue:333]: Is used differently each time. (sometimes with CategoryChip, sometimes without)
So I'm not sure if this even makes sense. Maybe as a component?

**Review by manuelmeister — APPROVED** [review]
Cool

File: frontend/src/components/activity/CopyActivityInfoDialog.vue [hunk 31]
```diff
@@ -0,0 +1,75 @@
...
+      <center>
+        <v-btn color="success" @click="requestClipboardAccess">
+          {{ $tc('components.activity.copyActivityInfoDialog.allow') }}
+        </v-btn>
+      </center>
+    </p>
+    <p v-if="clipboardReadState === 'granted'">
+      {{ $tc('components.activity.copyActivityInfoDialog.granted') }}
+    </p>
+    <p v-if="clipboardReadState === 'denied'">
```
Comment (BacLuc) [inline frontend/src/components/activity/CopyActivityInfoDialog.vue:26]: minor:
Could we keep the button to request access also in the denied case?
On some browsers (e.g. mobile) the button to grant the permission is not that easy to find.

File: frontend/src/components/program/DialogActivityCreate.vue [hunk 32]
```diff
@@ -16,21 +16,107 @@
...
+              {{ copyActivitySource.title }}
+            </v-list-item-title>
+            <v-list-item-subtitle>
+              {{ copyActivitySource.camp().title }}
+            </v-list-item-subtitle>
+          </v-list-item-content>
+          <v-list-item-action>
+            <e-checkbox
+              v-model="copyContent"
+              :label="$tc('components.program.dialogActivityCreate.copyActivityContent')"
```
Comment (BacLuc) [inline frontend/src/components/program/DialogActivityCreate.vue:68]: minor:
The label text content is a little long on mobile and hides the activity title.
But you see it in the title field and in the category field

File: frontend/src/components/program/DialogActivityCreate.vue [hunk 33]
```diff
@@ -61,15 +198,94 @@ export default {
...
+        },
+        () => {
+          this.copyActivitySourceUrlLoading = false
+        }
+      )
+    },
   },
   methods: {
+    refreshCopyActivitySource() {
+      navigator.permissions.query({ name: 'clipboard-read' }).then(
```
Comment (BacLuc) [inline frontend/src/components/program/DialogActivityCreate.vue:253]: minor:
On mobile duckduckgo browser (and probably firefox) this leads to an error.
https://ecamp.sentry.io/issues/4633431420
Maybe change the onRejected handler to a catch which catches the whole chain?

File: frontend/src/components/program/DialogActivityCreate.vue [hunk 34]
```diff
@@ -61,15 +198,94 @@ export default {
...
+        (p) => {
+          this.$set(this, 'clipboardPermission', p.state)
+          this.$set(this, 'copyActivitySource', null)
+
+          if (p.state == 'granted') {
+            navigator.clipboard.readText().then((url) => {
+              this.getCopyActivitySource(url).then((activityProxy) => {
+                activityProxy._meta.load.then((activity) =>
+                  this.$set(this, 'copyActivitySource', activity)
```
Comment (BacLuc) [inline frontend/src/components/program/DialogActivityCreate.vue]: minor:
the experience is not the same if the user gave clipboard access or if he did not.
If you detect an activity in the clipboard, the checkbox to copy the the activity content is not set.
If the user manually pastes the url, the checkbox to copy the activity content is set.

Is this on purpose?

**Review by BacLuc — APPROVED** [review]

File: frontend/src/components/activity/ScheduleEntry.vue [hunk 35]
```diff
@@ -117,6 +117,18 @@ Displays a single scheduleEntry
...
 
+          <v-list-item @click="copyUrlToClipboard">
+            <v-list-item-icon>
+              <v-icon>mdi-content-copy</v-icon>
+            </v-list-item-icon>
+            <v-list-item-title>
+              {{ $tc('components.activity.scheduleEntry.copyScheduleEntry') }}
+            </v-list-item-title>
+          </v-list-item>
+          <copy-activity-info-dialog ref="copyInfoDialog" />
```
Comment (manuelmeister) [inline frontend/src/components/activity/ScheduleEntry.vue]:
```suggestion
          <CopyActivityInfoDialog ref="copyInfoDialog" />
```

As it is a special self made component with a single use it should follow the PascalCase.

File: frontend/src/components/program/DialogActivityCreate.vue (hunk 33 above)
Comment (manuelmeister) [reply frontend/src/components/program/DialogActivityCreate.vue:253]: I think you have to add a `.catch()` after the `.then(…)`.

**Review by manuelmeister — CHANGES_REQUESTED** [review]

**Comment by manuelmeister** [comment]
I created a PR to resolve the leftover issues: https://github.com/pmattmann/ecamp3/pull/7

**Comment by BacLuc** [comment]
You need to merge devel again after #4136

**Review by manuelmeister — APPROVED** [review]
